Add validation for credentials in Openstack Cloud Provider

aufi opened the pull request:
https://bugzilla.redhat.com/show_bug.cgi?id=1213921

chessbyte commented:
@blomquisg please review

blomquisg commented on app/models/manageiq/providers/openstack/cloud_manager.rb:
```diff
@@ -31,6 +31,8 @@ class ManageIQ::Providers::Openstack::CloudManager < EmsCloud
 
   include ManageIQ::Providers::Openstack::ManagerMixin
 
+  validates :authentications, :presence => {:message => 'You must enter credentials.'}
```
Hrm, if we're going to do this, then it probably makes more sense to do this at either the [`cloud_manager`](https://github.com/ManageIQ/manageiq/blob/3447f98fed91bb8d015f6ffc9755c9bd7d27e47a/app/models/manageiq/providers/cloud_manager.rb) level or [`base_manager`](https://github.com/ManageIQ/manageiq/blob/3447f98fed91bb8d015f6ffc9755c9bd7d27e47a/app/models/manageiq/providers/base_manager.rb) level.

Otherwise, we're going to get different behavior across different providers for credentials requirements.

/cc @Fryguy @juliancheal @jrafanie 

Thoughts?

juliancheal replied:
Seems to work well in `cloud_manager`, but then don't we also need it in `infra_manager`.
I'd prefer `base_manager`.
Also it seems Foreman already requires credentials :)

jrafanie replied:
I'm fine with using `base_manager`... how would a subclass disable such a validation?  Via a flag at `base_manager` that a subclass could override as false?

aufi commented:
I moved the validation to Manageiq::Providers::Basemanager and added custom validation method which can be redefined in subclasses to skip the validation.
Does it look acceptable? (if so, I would add commit with validation skip in other providers for now)

chessbyte commented:
@blomquisg please review
@matthewd please comment on https://github.com/ManageIQ/manageiq/pull/6124#issuecomment-174935187

blomquisg commented on app/models/manageiq/providers/base_manager.rb:
```diff
@@ -2,6 +2,8 @@ module ManageIQ::Providers
   class BaseManager < ExtManagementSystem
     require_nested :Refresher
 
+    validate :require_credentials
```
@juliancheal does this impact you at all?

juliancheal replied:
Checked this against the latest master (with delegation to endpoints) and it seems fine for now.

blomquisg commented on spec/factories/ext_management_system.rb:
```diff
@@ -89,13 +89,16 @@
+    after :build do |x|
+      x.authentications << FactoryGirl.create(:authentication, :userid => "admin", :password => "123456789")
+      x.authentications << FactoryGirl.create(:authentication, :userid => "user", :password => "abcdefgh",
+                                              :authtype => "amqp")
```
I wonder if it makes sense to keep the authentications separate from the base `ems_openstack` factory.  @Ladas, thoughts?

Ladas replied:
I guess it could stay in the factory :ems_openstack_with_authentication, so it's consistent with the other factories. Maybe there is a good reason for moving it? :-)

aufi replied:
Authentications were moved back to :ems_openstack_with_authentication.

matthewd commented on app/models/manageiq/providers/base_manager.rb:
```diff
@@ -2,6 +2,8 @@ module ManageIQ::Providers
   class BaseManager < ExtManagementSystem
```
This class is a bit of a wart we should clean up.. but for now, the below changes should probably go on `ExtManagementSystem` instead.

juliancheal replied:
:+1: `ExtManagementSystem` might be a better fit.

aufi replied:
Makes sense, moved.

matthewd commented on app/models/manageiq/providers/base_manager.rb:
```diff
@@ -17,5 +19,11 @@ def ext_management_system
+    private
+
+    def require_credentials
+      errors.add(:authentications, 'You must enter credentials.') if missing_credentials?
```
I feel like validations that require child objects to exist can be potentially problematic.. but maybe this is okay?

Ladas replied:
We are already doing that, we are validating hostname, that was moved to authentications. So I guess it's ok?

The check right now is, that the default authentication exists and it has a hostname. This check should just add checking credentials on this default authentication, which are marked as required in the UI, so it should be ok.

matthewd replied:
Oh.. do we want `validates_associated`, then?

aufi replied:
A methods in EmsCloudController does not add "empty" authentication to EMS. In case credentials were not filled, EMS gets no empty authentication, but none. In that case validates_associated does not fail.
I am thinking about adjusting controller method with a transient attribute on ems and checking it in custom validation, will update PR.

blomquisg commented:
I'd rather do this validation, where it's just checking for the presence of field values, in javascript.  But, @dclarizio said that we can't move this to javascript until the add/edit provider pages are all moved to angular.

So, I'm going to close this PR and suggest that we do this in javascript when we have all the add/edit provider pages converted to angular.
